### PR TITLE
Emmarothwell1/restricted function space

### PR DESCRIPTION
**Pull request by dham**
Add support for function space labels in order to support symbolically different function spaces on the same element and domain. An example are restricted function spaces.

### CODE REVIEW

File: ufl/functionspace.py [hunk 1]
```diff
@@ -50,11 +50,16 @@ def __init__(self, domain, element):
             else:
                 if element.cell != domain_cell:
                     raise ValueError("Non-matching cell of finite element and domain.")
-
+
```
Comment (dham) [inline ufl/functionspace.py:53]:
```suggestion

```

File: ufl/functionspace.py [hunk 2]
```diff
@@ -80,6 +85,7 @@ def _ufl_hash_data_(self, name=None):
         name = name or "BaseFunctionSpace"
         domain = self.ufl_domain()
         element = self.ufl_element()
+        bdata = self.label()
```
Comment (dham) [inline ufl/functionspace.py:88]: I guess we don't need this variable, just directly use `self.label` below.

File: ufl/functionspace.py [hunk 3]
```diff
@@ -88,13 +94,14 @@ def _ufl_hash_data_(self, name=None):
...
             edata = element._ufl_hash_data_()
-        return (name, ddata, edata)
+        return (name, ddata, edata, bdata)
 
     def _ufl_signature_data_(self, renumbering, name=None):
         """UFL signature data."""
         name = name or "BaseFunctionSpace"
         domain = self.ufl_domain()
         element = self.ufl_element()
+        bdata = self.label()
```
Comment (dham) [inline ufl/functionspace.py:104]: as above.

File: ufl/functionspace.py [hunk 4]
```diff
@@ -50,11 +50,15 @@ def __init__(self, domain, element):
...
                 if element.cell != domain_cell:
                     raise ValueError("Non-matching cell of finite element and domain.")
-
         AbstractFunctionSpace.__init__(self)
+        self._label = label
         self._ufl_domain = domain
         self._ufl_element = element
 
+    def label(self):
+        """Return label of boundary domains to differentiate restricted and unrestricted"""
```
Comment (mscroggs) [inline ufl/functionspace.py:59]:
```suggestion
        """Return label of boundary domains to differentiate restricted and unrestricted."""
```

**Pull request merged**